// @unipasswallet/utils/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MultiCallAddress = exports.MultiCallInterface = exports.ModuleHookEIP4337WalletInterface = exports.ModuleGuestInterface = exports.ModuleMainUpgradableInterface = exports.ModuleMainInterface = exports.SingletonFactoryInterface = exports.SingletonFactoryAddress = exports.CreationCode = void 0;
const utils_1 = require("ethers/lib/utils");
const abi_1 = require("@unipasswallet/abi");
exports.CreationCode = "0x603a600e3d39601a805130553df3363d3d373d3d3d363d30545af43d82803e903d91601857fd5bf3";
exports.SingletonFactoryAddress = "0xce0042b868300000d44a59004da54a005ffdcf9f";
exports.SingletonFactoryInterface = new utils_1.Interface(abi_1.singletonFactory.abi);
exports.ModuleMainInterface = new utils_1.Interface(abi_1.moduleMain.abi);
exports.ModuleMainUpgradableInterface = new utils_1.Interface(abi_1.moduleMainUpgradable.abi);
exports.ModuleGuestInterface = new utils_1.Interface(abi_1.moduleGuest.abi);
exports.ModuleHookEIP4337WalletInterface = new utils_1.Interface(abi_1.moduleHookEIP4337Wallet.abi);
exports.MultiCallInterface = new utils_1.Interface(abi_1.multiCall.abi);
exports.MultiCallAddress = "0x175d02d277eac0838af14D09bf59f11B365BAB42";
